src/system/sargon_os/sargon_os.rs

```diff
@@ -0,0 +1,478 @@
+use std::sync::Once;
+
+use crate::prelude::*;
+
+/// The Sargon "Operating System" is the root "manager" of the Sargon library
+/// which holds an in-memory Profile and a collection of "clients" which are
+/// created from "drivers" which the hosts (iOS/Android wallets) "installs"
+/// during app launch, enabling the  Sargon "Operating System" to e.g read/write
+/// to secure storage and make use of the network connection of the iPhone/Android
+/// phone.
+#[derive(Debug, uniffi::Object)]
+pub struct SargonOS {
+    pub(crate) profile_holder: ProfileHolder,
+    pub(crate) clients: Clients,
+}
+
+/// So that we do not have to go through `self.clients`,
+/// but can use e.g. `self.secure_storage` directly.
+impl Deref for SargonOS {
+    type Target = Clients;
+
+    fn deref(&self) -> &Self::Target {
+        &self.clients
+    }
+}
+
+#[uniffi::export]
+impl SargonOS {
+    #[uniffi::constructor]
+    pub async fn boot(bios: Arc<Bios>) -> Result<Arc<Self>> {
+        Self::boot_with_bdfs(bios, None).await
+    }
+}
+
+impl SargonOS {
+    pub async fn boot_with_bdfs(
+        bios: Arc<Bios>,
+        bdfs_mnemonic: Option<MnemonicWithPassphrase>,
+    ) -> Result<Arc<Self>> {
+        let clients = Clients::new(bios);
+
+        let sargon_info = SargonBuildInformation::get();
+        let version = sargon_info.sargon_version;
+        let ret_version = sargon_info.dependencies.radix_engine_toolkit;
+        info!("Booting SargonOS {} (RET: {})", &version, &ret_version);
+        let host_info = clients.host.summary().await;
+        info!("Host: {}", host_info);
+
+        let secure_storage = &clients.secure_storage;
+
+        if let Some(loaded) = secure_storage.load_active_profile().await? {
+            info!("Loaded saved profile {}", &loaded.header);
+            let is_owner = Self::check_is_allowed_to_update_provided_profile(
+                &clients, &loaded, false,
+            )
+            .await?;
+
+            if !is_owner {
+                warn!("Loaded saved profile was last used on another device, will continue booting OS, but will unable to update Profile.");
+            }
+
+            Ok(Arc::new(Self {
+                clients,
+                profile_holder: ProfileHolder::new(loaded),
+            }))
+        } else {
+            info!("No saved profile found, creating a new one...");
+            let (profile, bdfs) =
+                Self::create_new_profile_with_bdfs(&clients, bdfs_mnemonic)
+                    .await?;
+
+            secure_storage.save_private_hd_factor_source(&bdfs).await?;
+
+            secure_storage
+                .save_profile_and_active_profile_id(&profile)
+                .await?;
+
+            info!("Saved new Profile and BDFS, finish booting SargonOS");
```

Comment:
Which part **specifically** do you think should move to `ProfileHolder`? `ProfileHolder` does **NOT** have the clients, nor do I think it should, it should only "hold the Profile". But yes we might wanna introduce one more layer though!

CURRENT (current state of this PR):
`SargonOS(HAS ALL clients) -> ProfileHolder(NO clients) -> Profile`

ONE MORE LAYER:
`SargonOS(HAS ALL clients) -> ProfileManager(HAS SOME clients) -> ProfileHolder(NO clients) -> Profile`

But maybe something that can be done later?

Reply:
I am also asking this in the context of what would be the standard, should only SargonOS operate with clients, or subcomponents can have clients?

I would add RadixConnectMobile which would need some clients.

As for having one more layer, ProfileHolder by itself doesn't seem quite useful, is there a reason I am missing as to why not have just ProfileManager that operates on the Profile?

Reply:
It has to do with Rust and UniFFI :) `SargonOS` cannot UniFFI export `*mut self` methods. So we cannot use `mut`. Thus SargonOS MUST have "interior mutability". But UniFFI cannot use `RefCell` (or `Cell`) so we MUST use `RwLock`. And working with that is .... a bit cumbersome, so that code I dont want in `SargonOS`, but it has to go *somewhere*, that is the purpose of `ProfileHolder`! To be able to **AT ALL** mutate a `Profile` and still work with UniFFI :) 

`RadixConnectMobile` is probably going to be a client! It might be a client which uses other clients - which would be a first. So maybe we should introduce one new "classifications"? Maybe "component"? So we call it `RCMComponent` and we say: "'Components' are allowed to have 'Clients'" - and then we call `ProfileManager` `ProfileManagerComponent`? 

I think that would work quite well?

Reply:
> RadixConnectMobile is probably going to be a client

Isn't client reserved to drivers? I don't expect Wallets to have to init a RadixConnectMobile client, so with your terminology it would be more of a component, similar to ProfileManager.

As Sargon would grow, we will have more and more components for the sake of encapsulation.

Probably `Module` would be a better term in the context of OS, but might be confusing in the context of Cargo :).

Reply:
Clients are "powered by" a driver yes. 

So lets aim for:
`ProfileComponent`
`RCMComponent`

And components have modules. `Client` are reference types, behind `Arc` so we can clone them, the same underlying instance is used. And we pass e.g. `HttpClient` into `RCMComponent` and possibly others.

If something needs to interact with profile, it should do so via `SargonOS` - which is the "coordinator of components". 

Maybe that would work?

Reply:
What is your idea with going via SargonOS?
Personally I would be ok with passing down ProfileComponent to other components as ref, or have a more specialised trait in between depending on the said component needs.

As for `Arc`, I think we can just use references directly, I mean the clients basically have a static lifetime in regards to SargonOS; the clients will never be deallocated until the SargonOS itself will be deallocated.

Reply:
I think it is a simpler model to understand, that wallets always go via SargonOS.

All this is quite complex, why I tried to come up with a mental model which helps - an "OS with drivers". 

Apart from that it is convenient to not have to keep many singletons around in wallets (RCM, GW?, OS): my idea is wallet only keep OS around (and due to UniFFFI impl details/limitations in Rust - also the EventBus singleton).

The ProfileHolder / ProfileManager(Component) would be dangerous to clone, since the RwLock might cause crash if we have multiple writers! Or writer and reader simultaneously. Having one RwLock prevents that from happening. These are "deep rust" where I'm no expert! But I think it can be dangerous to let many components access Profile. Having one OS, coordinating Profile read/write access protects against that, thus all components that might care about Profile should be children / hierarchically below OS.

Maybe best sync this in a video call?

Reply:
With having the Wallets to go through SargonOS always I mostly agree, I am not against it.

I mostly was referring to this:
> If something needs to interact with profile, it should do so via SargonOS - which is the "coordinator of components".

Where `something` can also be other component that needs to interact with Profile.

Reply:
right, so I think the solution is the structure implemented in this PR, where:

* `Host` (Wallet apps) "requests" `SargonOS` to add (save) a created but not yet saved account
* SargonOS 
  - Checks ownership by loading `DeviceInfo` using `SecureStorageClient` (using `SecureStorageDriver`) and comparing against `ProfileHolder`s held Profile
  - Adds account into profile held by `ProfileHolder`
  - Persist updated profile using `SecureStorageClient`
  - Tells `EventBusClient` (using `EventBusDriver`) to emit `EventNotification::ProfileModified(EventProfileModified::AccountAdded)`

I think that is what we can do and should do. SargonOS coordinates reads/writes from/to `ProfileHolder` and uses clients to perform tasks. I think it should be applicable to all our needed features?

Can you think of any scenario where that setup/structure does not work?

Reply:
By applying this to all possible components that SargonOS would make it quite big. I see that in your structure you are biased towards SargonOS managing Profile, so it feel more like ProfileOS ;), so you have all of the Profile related logic in SargonOS. If we are to use the same approach for RadixConnectMobile, 90% of its code will have to be actually in SargonOS. 

What you have described above, it should ideally be encapsulated in a ProfileManagerComponent. What are your concerns about having such component while SargonOS acts as a composition root and as a facade to its internal subcomponents?

Reply:
No concerns other than; what happens if FooComponent of BarComponent needs to access profile?

Reply:
I meant `FooComponent OR BarComponent` :)

Reply:
The problem is ensuring that we never do multiple reads or one read and one write simultaneuolsy on the `RwLock` in `ProfileHolder` (`ProfileManagerComponent`). As long as we have a plan to prevent that - knowingly prevent that - which is what `actor`s in Swift are for, which protects at compile and runtime :). 

We can switch from `try_read` to `read` on the `RwLock` which will wait for the ReadGuard, maybe that is ok? But yeah we will just have to explore it a little bit.

Reply:
Totally agree. 
But this problem, at least in my understanding, does not appear suddenly because other Components now interact with the profile from within, it will happen at any point in time there is a concurrent read/write, which can happen even if we have everything in SargonOS and we do forbid components to directly read/write the Profile. After all the Wallet could call SargonOS API from different concurrent contexts, like doing some change on the Profile and then background cloud backup kicks in.

Reply:
mm maybe we just need to be a bit lenient and pragmatic and solve issues when we face them, instead of assuming they will happen. And if needed, we can protect against concurrent access by putting a "holder" around `SargonOS` singleton in Swift / Kotlin, which in Swift would be an `actor` which at least would protect against concurrent read / write requests **originating from Host**.

Reply:
but wouldn't that be too much to put a holder around whole SargonOS? does it mean that a potential RadixConnectMobile request might be blocked by some unrelated Profile change?

Profile is quite likely to be accessed concurrently with MFA(also potentially with cloud backups done in background), so we should start thinking about it. An option seems to be using actor model in Sargon directly with [actix](https://crates.io/crates/actix), although it is likely to make the read/writes async.

Reply:
true, Im not sure what we wanna do. if we can find truely distinct components, where RadixConnectMobile does not use Profile **at all** then it is easily built. We need to give it some thought. And while giving it thought, we need to eval "current state" and "after having moved to Workspace and only One crate uses UniFFI" scenarios. The former prevents us from using `actix` actors, since they are not UniFFI compatible (and probably very hard to be made UniFFI compatible).

Reply:
What is the reason for actors to be exposed through UniFFI? You say that this:
```
#[derive(Debug, uniffi::Object)]
struct SargonOS {
   profile_manager: Actor
}
```
cannot be exposed as an object?

Reply:
I was wrong! We can! Works with `uniffi::Object` indeed! - But does not work with `uniffi::Record`... I'm too used to thinking about records...

Reply:
Amazing then! We can pursue using Actors, hopefully it works smoothly 🤞